src/CoreWCF.Primitives/src/CoreWCF/Channels/SslStreamSecurityUpgradeProvider.cs

```diff
@@ -297,8 +297,8 @@ protected override async Task<(Stream, SecurityMessageProperty)> OnAcceptUpgrade
                 _clientCertificate = certificate2;
                 try
                 {
-                    SecurityToken token = new X509SecurityToken(certificate2, false);
-                    ReadOnlyCollection<IAuthorizationPolicy> authorizationPolicies = _parent.ClientCertificateAuthenticator.ValidateToken(token);
+                    SecurityToken token = new X509SecurityToken(certificate2, true);
+                    ReadOnlyCollection<IAuthorizationPolicy> authorizationPolicies = _parent.ClientCertificateAuthenticator.ValidateTokenAsync(token).AsTask().GetAwaiter().GetResult();
                     _clientSecurity = new SecurityMessageProperty
```

Comment:
We want to avoid allocation if possible, which is why we're using ValueTask instead of Task. Do something like this instead which will not allocate if using the default built-in authenticator:
```c#
ReadOnlyCollection<IAuthorizationPolicy> authorizationPolicies;
var validationValueTask = _parent.ClientCertificateAuthenticator.ValidateTokenAsync(token);
if (validationValueTask.IsCompleted)
{
    authorizationPolicies = validationValueTask.Result; // Zero allocation
}
else
{
    authorizationPolicies = validationValueTask.AsTask().GetAwaiter().GetResult();
}
```

This gives the benefits of zero allocation synchronous completion while still supporting async completion.

Reply:
done